File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-4.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func lowestCommonAncestor(root *TreeNode, p *TreeNode, q *TreeNode) *TreeNode {
	if root == nil{
		return root
	}
	rootVal := root.Val
	pVal := p.Val 
	qVal := q.Val
	ans := root

	if pVal <= rootVal && qVal >= rootVal || pVal >= rootVal && qVal <= rootVal{
		return ans
	}
	if pVal < rootVal && qVal < rootVal{
		ans = lowestCommonAncestor(root.Left, p, q)
	}else{
		ans = lowestCommonAncestor(root.Right, p, q)
	}
	return ans
}

File: Data Structures & Algorithms/valid-binary-search-tree/submission-4.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    void dfs(TreeNode*root, bool &ans,int min,int max)
    {
        if(!root || !ans)
        {
            return;
        }
        if(root->val <= min || root->val >= max)
        {
            ans = false;
            return;
        }

        dfs(root->left,ans,min,root->val);
        dfs(root->right,ans,root->val,max);
        return;
        
        
    }
    bool isValidBST(TreeNode* root) {
        bool ans = true;
        dfs(root,ans,INT_MIN,INT_MAX);
        return ans;
        
        
    }
};
